ratpack-micrometer-metrics/src/main/java/ratpack/micrometer/metrics/HandlerTags.java

```diff
@@ -0,0 +1,201 @@
+    HandlerTags.method(request), HandlerTags.uri(request, null)
+  );
+
+  public static final BiFunction<Context, Throwable, Tags> RECOMMENDED_HANDLER = (context, exception) -> Tags.of(
+    HandlerTags.method(context.getRequest()), HandlerTags.uri(context.getRequest(), context.getResponse()),
+    HandlerTags.exception(exception), HandlerTags.status(context.getResponse()), HandlerTags.outcome(context.getResponse()));
+
+  private static final Tag URI_NOT_FOUND = Tag.of("uri", "NOT_FOUND");
+
+  private static final Tag URI_REDIRECTION = Tag.of("uri", "REDIRECTION");
+
+  private static final Tag URI_ROOT = Tag.of("uri", "root");
+
+  private static final Tag URI_UNKNOWN = Tag.of("uri", "UNKNOWN");
+
+  private static final Tag EXCEPTION_NONE = Tag.of("exception", "None");
+
+  private static final Tag STATUS_UNKNOWN = Tag.of("status", "UNKNOWN");
+
+  private static final Tag METHOD_UNKNOWN = Tag.of("method", "UNKNOWN");
+
+  private static final Pattern LEADING_SLASH_PATTERN = Pattern.compile("^/");
+
+  private static final Pattern TRAILING_SLASH_PATTERN = Pattern.compile("/$");
+
+  private static final Pattern MULTIPLE_SLASH_PATTERN = Pattern.compile("//+");
+
+  private HandlerTags() {
+  }
+
+  /**
+   * Creates a {@code method} tag based on the {@link Request#getMethod()
+   * method} of the given {@code request}.
+   * @param request the request
+   * @return the method tag whose value is a capitalized method (e.g. GET).
+   */
+  public static Tag method(Request request) {
+    return (request != null) ? Tag.of("method", request.getMethod().getName()) : METHOD_UNKNOWN;
+  }
+
+  /**
+   * Creates a {@code status} tag based on the status of the given {@code response}.
+   * @param response the HTTP response
+   * @return the status tag derived from the status of the response
+   */
+  public static Tag status(Response response) {
+    return (response != null) ? Tag.of("status", Integer.toString(response.getStatus().getCode())) : STATUS_UNKNOWN;
+  }
+
+  /**
+   * Creates a {@code uri} tag based on the URI of the given {@code request}. Falling back to {@code REDIRECTION} for 3xx responses,
+   * {@code NOT_FOUND} for 404 responses, {@code root} for requests with no path info, and {@code UNKNOWN} for all other requests.
+   * @param request the request
+   * @param response the response
+   * @return the uri tag derived from the request
+   */
+  public static Tag uri(Request request, @Nullable Response response) {
+    if (request != null) {
+      if (response != null) {
+        Status status = response.getStatus();
+        if (status != null) {
+          if (status.is3xx()) {
+            return URI_REDIRECTION;
+          }
+          if (status == Status.NOT_FOUND) {
+            return URI_NOT_FOUND;
+          }
+        }
+      }
+      String pathInfo = getPathInfo(request);
+      if (pathInfo.isEmpty()) {
+        return URI_ROOT;
+      } else {
+        return Tag.of("uri", pathInfo);
```

Comment:
won't this generate high cardinality metrics when path variables are used in requests?

Reply:
Agreed - we need a way to either 1) generate tags based on path tokens or 2) make this configurable. @smcadams86 can you share the code you have that uses path tokens instead of resolved values?

Reply:
I grabbed the `description` from the `pathBinding` object: https://github.com/ratpack/ratpack/blob/master/ratpack-core/src/main/java/ratpack/path/PathBinding.java#L88

Reply:
Yes, thank you @smcadams86! Meant to get back to this case. Appreciate the code sample as well.

Reply:
Resolved in 7e0b36c.